metrics/resource_view.go

```diff
@@ -51,14 +54,51 @@ type meters struct {
 	meters  map[string]*meterExporter
 	factory ResourceExporterFactory
 	lock    sync.Mutex
+	clock   clock.Clock
+	ticker  clock.Ticker
 }
 
 // Lock regime: lock allMeters before resourceViews. The critical path is in
 // optionForResource, which must lock allMeters, but only needs to lock
 // resourceViews if a new meter needs to be created.
-var resourceViews = storedViews{}
-var allMeters = meters{
-	meters: map[string]*meterExporter{"": &defaultMeter},
+var (
+	resourceViews = storedViews{}
+	allMeters     = meters{
+		meters: map[string]*meterExporter{"": &defaultMeter},
+		clock:  cleanupClock,
+	}
+
+	cleanupClock               = clock.Clock(clock.RealClock{}) // Var needed to avoid allMeters->defaultMeter->allMeters.clock cycle
+	cleanupOnce                = new(sync.Once)
+	meterExporterScanFrequency = 1 * time.Minute
+	maxMeterExporterAge        = 10 * time.Minute
+)
+
+func cleanup() {
+	log.Printf("in cleanup")
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+	expiryCutoff := allMeters.clock.Now().Add(-1 * maxMeterExporterAge)
+	for key, meter := range allMeters.meters {
+		if key != "" && meter.t.Before(expiryCutoff) {
+			flushGivenExporter(meter.e)
+			delete(allMeters.meters, key)
+		}
+	}
+
+}
+
+func cleanupLoop() {
+	for range allMeters.ticker.C() {
+		cleanup()
+	}
+}
+
+func startCleanup() {
```

Comment:
should this just be in `init()` then locking is not necessary?

Reply:
In resource_view.go, there is a ClearMetersForTest.  I felt like this was close enough in its intent to being a reset function, so I thought I would keep this around so the whole library can be reinitialized.

The only test that currently needs the clock is TestAllMeterExpiration, which is new.  If we don't allow swapping the clock out after initialization (that is, allow swapping a running time.Ticker for a new one based on a fake clock), then this test function needs to run before anything else to set the clock.  This is error prone as any new test could break this ordering.